### PR TITLE
adds a tabcontroller

### LINKED ISSUE #191: code cleanup
Write a tabbed component class, and refactor the dart/html/css and console/documentation tabbed area to use that. It'll probably live here: https://github.com/dart-lang/dart-pad/blob/master/lib/elements/elements.dart.

The `Tabbed` object should contains `Tab`s, and you should be able to get and get the selected tab, and listen for selected tab changes.

### PR DESCRIPTION
**Pull request by kasperpeulen**
not completely sure about the names, but it seems to all work fine at least

fix #191

### CODE REVIEW

**Comment by devoncarew** [comment]
Cool, taking a look now.

There's a build failure related to an unused getter: `[info] The getter '_resultTab' is not used at lib/playground.dart, line 55.`

File: lib/elements/elements.dart [hunk 1]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
   // TODO:
   Stream get onChanged => null;
 }
+
+class TabController {
+  List<TabElement> tabElements;
```
Comment (devoncarew) [inline lib/elements/elements.dart]: `final`?

File: lib/elements/elements.dart [hunk 2]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
...
+
+  TabElement _selectedTab;
+
+  TabController({this.tabElements}) {
```
Comment (devoncarew) [inline lib/elements/elements.dart]: `this.tabElements: []`

File: lib/elements/elements.dart [hunk 3]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
...
+    });
+  }
+
+  void selectTab(String tabName) {
+    selectedTab = tabElements.firstWhere((t) => t.name == tabName);
+  }
+}
+
+class TabElement extends DElement {
+  String name;
```
Comment (devoncarew) [inline lib/elements/elements.dart]: `final`?

File: lib/elements/elements.dart [hunk 4]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
...
+
+  Function onSelect;
```
Comment (devoncarew) [inline lib/elements/elements.dart]: `final`?

File: lib/elements/elements.dart [hunk 5]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
...
+    } else {
+      tabElements = [];
+    }
+  }
+
+  TabElement get selectedTab => tabElements.firstWhere((tab) => tab.hasAttr("selected"));
+
+  set selectedTab(TabElement s) {
+    _selectedTab = s;
+    tabElements.where((tab) => tab != _selectedTab).forEach((tab) {
```
Comment (devoncarew) [inline lib/elements/elements.dart]: This is equivalent (whichever you prefer :) )

```
for (Tabelement tab in tabElements) {
  tab.toggleAttr('selected', tab == _selectedTab);
}
```

File: lib/elements/elements.dart [hunk 6]
```diff
@@ -505,3 +505,49 @@ class _ElementTextProperty implements Property {
...
+
+  void registerTab(TabElement tab) {
+    tabElements.add(tab);
+    tab.onClick.listen((_) {
+      selectedTab = tab;
+      tab.onSelect();
+    });
+  }
+
+  void selectTab(String tabName) {
```
Comment (devoncarew) [inline lib/elements/elements.dart]: Perhaps note in a dartdoc comment that this method will throw if the `tabName` is not the name of a current tab.

**Comment by devoncarew** [comment]
This is awesome! This cleans up the code in `playground.dart` and gives us a nice OO wrapper around tabs.

File: lib/elements/elements.dart (hunk 1 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: I think this is not possible, as registerTab adds tabs to this list.

File: lib/elements/elements.dart (hunk 2 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: If I do this, I get the error that default values of optional parameters must be constant.

File: lib/elements/elements.dart (hunk 1 above)
Comment (devoncarew) [inline lib/elements/elements.dart]: The variable can be `final` - you can't assign a new list to it - but you can still modify the referenced list itself.

File: lib/elements/elements.dart (hunk 2 above)
Comment (devoncarew) [inline lib/elements/elements.dart]: OK, nevermind then :)

File: lib/elements/elements.dart (hunk 4 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: I do the onSelect after initialization:

```
      controller.registerTab(new TabElement(querySelector('#${name}tab'), name: name)
        ..onSelect = () {
          Element issuesElement = querySelector('#issues');
          if (name != "dart") {
            issuesElement.style.display = "none";
          } else {
            issuesElement.style.display = "block";
          }
          ga.sendEvent('edit', name);
          _context.switchTo(name);
        }
      );
```

To make it more readable, I could change that, and make this field final then ?

File: lib/elements/elements.dart (hunk 5 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: ah, neat !

File: lib/elements/elements.dart (hunk 3 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: done

File: lib/elements/elements.dart (hunk 6 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: done

File: lib/elements/elements.dart (hunk 4 above)
Comment (kasperpeulen) [inline lib/elements/elements.dart]: okay, I did add a final here as well

**Comment by devoncarew** [comment]
Great, lgtm!

**Pull request merged**